feat: add pod start time panel

luciano-buono opened the pull request:
### Summary

JIRA: [DART-1473](https://saritasa.atlassian.net/browse/DART-1473)

Added new panel for Generic service metrics dashboard

Tested in ATS-staging cluster: https://grafana.staging.div25.com/d/qqsCbY5Zz/generic-service-metrics?orgId=1&refresh=10s&from=now-5m&to=now

Threshold line defined in 60s

Related PR: ?? (not created yet)
![image](https://github.com/user-attachments/assets/8b13a390-e4a6-4b8b-8f70-01e5aafdcd75)

![image](https://github.com/user-attachments/assets/547fc69c-557c-4cfe-bf81-6eb5e46f2830)

I created a new row for it, but we could add it to another row to better use the space. What do you think?


[DART-1473]: https://saritasa.atlassian.net/browse/DART-1473?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

luciano-buono committed:
feat: add pod start time panel

luciano-buono committed:
feat: changelog

luciano-buono committed:
fix: lower threshold line

luciano-buono commented:
![image](https://github.com/user-attachments/assets/88fb05ae-f714-45b5-abcf-d775b4a4b748)

I'm still debugging edge cases where the pods were restarted and we end up with very large (wrong) numbers for weeks:

dmitry-mightydevops commented:
@luciano-buono 

### we don't need **Time Series** panel for this data.

Display it as stat panel on top of the page
![image](https://github.com/user-attachments/assets/f1f6a658-394f-41fa-97be-b426fb4a273f)

So each panel should be the deployment name and aggregated data across all pods of the deployment.

### we :arrow_right: `need` to aggregate data by deployment (owner of the pod)

Right now you display data by pods, I am interested how fast the deployment gets ready to serve the traffic

I tested on usummit and the deployment data is not listed in labels
![image](https://github.com/user-attachments/assets/7dda3ad3-38e6-4131-82da-e29575f9a8c6)

Technically, this data should be there:
https://github.com/saritasa-nest/saritasa-devops-boilerplates/blob/5d0f9d257414217e6ebc6f5481f7534523c07798/kubernetes/prod/argocd/apps/templates/prometheus.yaml#L40-L42 (but it's not)

:sos: so let's fix it first. Don't join to other metrics (as it's slow) to get deployment name.


### The formula

Does not look correct to me

```
max(kube_pod_status_ready_time{} - kube_pod_created{}) by (pod, namespace)
  unless
  sum by (pod, namespace) (kube_pod_container_status_restarts_total > 0)
```

As kube_pod_created remains static (until the new deployment revision restarts the PODs) and OOMs will put bad data into results.

You should use something like

```
sum by (pod, namespace) (kube_pod_status_container_ready_time{namespace="prod", pod="keconnect-79864f958f-4qwrt"})
-
sum by (pod, namespace) (kube_pod_container_state_started{namespace="prod", pod="keconnect-79864f958f-4qwrt"})
```

and convert to actual container data (lowest level), rather than higher:
- pod
- replicase
- deployment/ss/ds etc

Also keep in mind this issue I noticed on usummit:
![image](https://github.com/user-attachments/assets/1a556f3f-f5f2-42ad-9e0e-898c8360f95e)

This may or may not affect you. Technically the following query should have worked, but it didn't for me due to labels mismatch as on the image.

```
kube_pod_status_container_ready_time{namespace="prod", pod="keconnect-79864f958f-4qwrt"}
-
kube_pod_container_state_started{namespace="prod", pod="keconnect-79864f958f-4qwrt"}
```

dmitry-mightydevops commented:
See in promql on [usummit](https://prometheus.teleport.usummitapp.net/graph?g0.expr=kube_pod_status_container_ready_time%7Bnamespace%3D%22prod%22%2C%20pod%3D%22keconnect-79864f958f-4qwrt%22%7D%0A-%0Akube_pod_container_state_started%7Bnamespace%3D%22prod%22%2C%20pod%3D%22keconnect-79864f958f-4qwrt%22%7D&g0.tab=1&g0.display_mode=lines&g0.show_exemplars=0&g0.range_input=1h&g1.expr=kube_pod_status_container_ready_time%7Bnamespace%3D%22prod%22%2C%20pod%3D%22keconnect-79864f958f-4qwrt%22%7D&g1.tab=1&g1.display_mode=lines&g1.show_exemplars=0&g1.range_input=1h&g2.expr=kube_pod_container_state_started%7Bnamespace%3D%22prod%22%2C%20pod%3D%22keconnect-79864f958f-4qwrt%22%7D&g2.tab=1&g2.display_mode=lines&g2.show_exemplars=0&g2.range_input=1h&g3.expr=sum%20by%20(pod%2C%20namespace)%20(kube_pod_status_container_ready_time%7Bnamespace%3D%22prod%22%2C%20pod%3D%22keconnect-79864f958f-4qwrt%22%7D)%0A-%0Asum%20by%20(pod%2C%20namespace)%20(kube_pod_container_state_started%7Bnamespace%3D%22prod%22%2C%20pod%3D%22keconnect-79864f958f-4qwrt%22%7D)&g3.tab=1&g3.display_mode=lines&g3.show_exemplars=0&g3.range_input=1h)

luciano-buono commented:
@dmitry-mightydevops 

```
sum by (pod, namespace) (kube_pod_status_container_ready_time{namespace="prod", pod="keconnect-79864f958f-4qwrt"})
-
sum by (pod, namespace) (kube_pod_container_state_started{namespace="prod", pod="keconnect-79864f958f-4qwrt"})
```

have similar problems.
![image](https://github.com/user-attachments/assets/5528faed-aed7-479c-ae04-424d7c86b95e)


In fact, you can check with kubectl all the timestamps and see that sometimes does not seem possible to get the real value:

```
❯ kgpo loki-canary-8jj8r -o yaml | yq .status
conditions:
  - lastProbeTime: null
    lastTransitionTime: "2024-08-05T17:28:22Z"
    status: "True"
    type: PodReadyToStartContainers
  - lastProbeTime: null
    lastTransitionTime: "2024-08-05T17:28:19Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2024-12-09T12:46:49Z"
    status: "True"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2024-12-09T12:46:49Z"
    status: "True"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2024-08-05T17:28:19Z"
    status: "True"
    type: PodScheduled
```

So these two seem to be always the same value: 
```
kube_pod_status_ready_time
kube_pod_status_container_ready_time
```

And this two also are the same:
```
kube_pod_container_state_started
kube_pod_created
```

Here you can see all labels which share the same values:  [PromQL ats-staging](https://prometheus.staging.div25.com/graph?g0.expr=max(kube_pod_status_ready_time%7Bpod%3D%22loki-canary-8jj8r%22%7D%20-%20kube_pod_created%7B%7D)%20by%20(pod%2C%20namespace)%0A%20%20unless%0A%20%20sum%20by%20(pod%2C%20namespace)%20(kube_pod_container_status_restarts_total%20%3E%200)&g0.tab=1&g0.display_mode=lines&g0.show_exemplars=0&g0.range_input=6h&g1.expr=kube_pod_status_container_ready_time%7Bpod%3D%22loki-canary-8jj8r%22%7D&g1.tab=1&g1.display_mode=lines&g1.show_exemplars=0&g1.range_input=1h&g2.expr=kube_pod_container_state_started%7Bpod%3D%22loki-canary-8jj8r%22%7D&g2.tab=1&g2.display_mode=lines&g2.show_exemplars=0&g2.range_input=1h&g3.expr=sum%20by%20(label_app_kubernetes_io_name%2C%20namespace%2C%20pod)%20(%0A%20%20(%0A%20%20%20%20sum%20by%20(pod%2C%20namespace)%20(kube_pod_status_container_ready_time%7Bpod%3D%22loki-canary-8jj8r%22%7D)%0A%20%20%20%20-%20sum%20by%20(pod%2C%20namespace)%20(kube_pod_container_state_started%7B%7D)%0A%20%20)%0A%20%20*%20on%20(pod)%20group_left(label_app_kubernetes_io_name)%20kube_pod_labels%0A)&g3.tab=1&g3.display_mode=lines&g3.show_exemplars=0&g3.range_input=1h&g4.expr=kube_pod_status_ready_time%7Bpod%3D%22loki-canary-8jj8r%22%7D&g4.tab=1&g4.display_mode=lines&g4.show_exemplars=0&g4.range_input=1h&g5.expr=kube_pod_created%7Bpod%3D%22loki-canary-8jj8r%22%7D&g5.tab=1&g5.display_mode=lines&g5.show_exemplars=0&g5.range_input=1h)

----
-----

> This may or may not affect you. Technically the following query should have worked, but it didn't for me due to labels mismatch as on the image.



Regading the label mismatch, this did not happen in ATS. But this is because we have different `metricLabelsAllowlist` in usummit compared to boilerplates.


Here you can see all the labels that are available for each metric:
https://github.com/kubernetes/kube-state-metrics/blob/main/docs/metrics/workload/pod-metrics.md

kube_pod_status_container_ready_time does not have the container label
kube_pod_container_state_started: it does


In usummit `kube_pod_status_container_ready_time` has this label as  `kube-state-metrics` because the Helm chart populates all the metrics defined there as their own.
(so it populates `wrong labels`)

Here it's explained: https://github.com/kubernetes/kube-state-metrics/issues/612

> This is the correct behavior with the configuration you have. Relabeling rules are applied to the discovered target, in this case kube-state-metrics is the target, it just happens to expose metrics about pods. I think what you’re looking for is the “kube_pod_labels” metric. Besides that I think you probably want to create an entirely separate scrape config just for kube-state-metrics that doesn’t relabel discovered labels in the way you have it.




----

Finally, regarding this:


> we ➡️ need to aggregate data by deployment (owner of the pod)
> Right now you display data by pods, I am interested how fast the deployment gets ready to serve the traffic

>  I tested on usummit and the deployment data is not listed in labels
image

> Technically, this data should be there:
> https://github.com/saritasa-nest/saritasa-devops-boilerplates/blob/5d0f9d257414217e6ebc6f5481f7534523c07798/kubernetes/prod/argocd/apps/templates/prometheus.yaml#L40-L42 (but it's not)

> 🆘 so let's fix it first. Don't join to other metrics (as it's slow) to get deployment name.


It does not seem possible to add a `deployment` label to the pods metric. Also, even if it's possible, it would have the same problem as above, it would just have the` Prometheus helm chart deployment` labels

dmitry-mightydevops commented:
@luciano-buono my method is actually working
the reason you get negative values is that you have double rows
![image](https://github.com/user-attachments/assets/785ccc19-4e67-4c98-9731-6f8f67916271)

modified the formula to use max and it worked

```
max by (pod, namespace) (kube_pod_status_container_ready_time{namespace="prod"})
-
max by (pod, namespace) (kube_pod_container_state_started{namespace="prod"})
```

![image](https://github.com/user-attachments/assets/7394f370-9241-4fe2-b126-a74648b5d8d7)

[promql](https://prometheus.teleport.usummitapp.net/graph?g0.expr=sum%20by%20(pod%2C%20namespace)%20(kube_pod_status_container_ready_time%7Bnamespace%3D%22prod%22%7D)%0A-%0Asum%20by%20(pod%2C%20namespace)%20(kube_pod_container_state_started%7Bnamespace%3D%22prod%22%7D)&g0.tab=1&g0.display_mode=lines&g0.show_exemplars=0&g0.range_input=1h&g1.expr=kube_pod_container_state_started%7Bnamespace%3D%22prod%22%2C%20pod%3D%22scrapyd-7b584774bb-5sw8d%22%7D&g1.tab=1&g1.display_mode=lines&g1.show_exemplars=0&g1.range_input=1h&g2.expr=kube_pod_status_container_ready_time%7Bnamespace%3D%22prod%22%2C%20pod%3D%22scrapyd-7b584774bb-5sw8d%22%7D&g2.tab=1&g2.display_mode=lines&g2.show_exemplars=0&g2.range_input=1h&g3.expr=max%20by%20(pod%2C%20namespace)%20(kube_pod_status_container_ready_time%7Bnamespace%3D%22prod%22%7D)%0A-%0Amax%20by%20(pod%2C%20namespace)%20(kube_pod_container_state_started%7Bnamespace%3D%22prod%22%7D)&g3.tab=1&g3.display_mode=lines&g3.show_exemplars=0&g3.range_input=1h)

dmitry-mightydevops reviewed (changes requested):
feedback above

dmitry-mightydevops commented:
@luciano-buono to accomodate multiple containers in the pod a proper would be to use max and min:

```
max by (pod, namespace) (kube_pod_status_container_ready_time{namespace="prod"})
-
min by (pod, namespace) (kube_pod_container_state_started{namespace="prod"})
```

luciano-buono committed:
feat: change panel for deployment start time

luciano-buono commented:
> to accomodate multiple containers in the pod a proper would be to use max and min:

@dmitry-mightydevops thanks for the tip. Modified

----
I kept the panels in the Generic Service Metrics dashboard and changed them to 'stat' type

![image](https://github.com/user-attachments/assets/b2a90906-a943-4e53-abd3-234ce85af393)
![image](https://github.com/user-attachments/assets/f1114b12-dc04-4f17-841d-c35834f3e179)

Check it here: https://grafana.staging.div25.com/d/qqsCbY5Zz/generic-service-metrics?orgId=1

luciano-buono committed:
feat: change to average values

luciano-buono committed:
feat: move deployment time block

luciano-buono committed:
feat: increase block size

kseniyashaydurova commented on general/generic-service-metrics.json:
```diff
@@ -31,6 +31,116 @@
+                    "uid": "prometheus"
+                },
+                "editorMode": "code",
+                "expr": "max(kube_pod_status_ready_time{namespace=\"$namespace\"}-kube_pod_created{namespace=\"$namespace\"}) by (pod)",
```
1. It seems it is a wrong formula?
![image](https://github.com/user-attachments/assets/42a2c333-430e-44c4-b3fa-98063a337960)

2. By the way, did you forgot to add ns for `kube_pod_container_state_started` metric in Grafana?

luciano-buono replied:
1. Why is it wrong? Do you mean syntax or logically?
2. It's unnecessary to add all the labels in both metrics since PromQL will only subtract the results that share all the exact labels. I avoided it to make it easier to read, but I can add it

luciano-buono replied:
Btw, for the namespaces variable in the dashboard, I filtered out all the namespaces except for the ones the devs are interested in:
![image](https://github.com/user-attachments/assets/fa764dce-5065-438d-a26a-7c0f409ee5b0)

But I think this will be a problem for rocks/cloud clusters, because we don't use staging|prod there, so we won't be able to select e.g.: `saritasa-chatbot` ns.
Should I remove this regex?

dmitry-mightydevops replied:
> Should I remove this regex?

@luciano-buono let's filter dynamically namespaces that have special tag
i.e. monitoring
if such tag is present in the namespace meta, then display it here.

luciano-buono replied:
Namespaces don't have any special tag that I'm aware of.

We could make some changes to add that but we would have to:
1. Modify the ArgoCD Application (in tekton-apps helm chart/boilerplates) to add a new label or annotation `monitoring=true` : 
https://argo-cd.readthedocs.io/en/stable/user-guide/sync-options/#namespace-metadata 
(and add `monitoring=false`, in all other namespaces)
2. Change the [kube-state--metrics](https://github.com/saritasa-nest/ats-dart-kubernetes-aws/blob/08cc2f5e1e3afaa612a03f7dc5872009cdef7c39/prod/argocd/apps/templates/prometheus.yaml#L41) to discover this new label/annotation
![image](https://github.com/user-attachments/assets/43b287c0-c59d-4ec7-b182-3ddc99caa10e)

3. After all that we should be able to filter here using: `monitoring = true`
![image](https://github.com/user-attachments/assets/788a3e8e-9a88-4671-adc3-f829cf87912b)
(Right now it's failing because we would need to add this annotation to all the metrics)

---
@dmitry-mightydevops , this seems too complex, was this your original idea?

dmitry-mightydevops replied:
Ok keep it simple

dmitry-mightydevops commented:
@luciano-buono pls work on @kseniyashaydurova feedback and merge it.

luciano-buono committed:
feat: remove regex in ns variable

luciano-buono commented:
@kseniyashaydurova I have removed the Regex and all the things tried https://github.com/saritasa-nest/saritasa-devops-grafana-dashboards/pull/14#discussion_r1931386525 to keep it simpler.

It's currently running here: https://grafana.staging.div25.com/d/qqsCbY5Zz/generic-service-metrics?orgId=1&refresh=10s
